### PR TITLE
remove time regex from value

### LINKED ISSUE #2858: Extreme slowdown due to use of std:regex on Windows (MSYS2 UCRT64)
#### **Describe the bug**
This is a follow-up of issue [2831](https://github.com/Exiv2/exiv2/issues/2831). In this case it concerns the std:regex used in the reading of IPTC date and time values.

#### **To Reproduce**
1. File made for testing by me 
exiv2-iptc-regex-utc64-1
2. Start UCRT64 MSYS2 gimp3 or self-built gimp master with up-to-date exiv2 package for UCRT64 (0.28.1-2) that includes the previous regex patch.
3. Open the above image and notice extreme slowdown and huge memory use.

#### **Expected behavior**
GIMP finishing opening the image in a short time.

#### **Desktop (please complete the following information):**
- OS and version: Windows 10 Home, 64-bit
- Exiv2 version and source: see above, but also self-built under MSYS2 UCRT64 with added printf statements to see where the slowdowns happened
- Compiler and version: gcc 13.2.0
- Compilation mode and/or compiler flags: Release

#### **Additional context**
By adding printf statements I observed the slowdowns this time in `src/value.cpp` :
1. in `DateValue::read` the line `static const std::regex reExtended(R"(^(\d{4})-(\d{2})-(\d{2}))");` possibly also the next line but I had to walk away for a while and when I came back it was in
2. `TimeValue::read`, slowdowns observed both in
	* `static const std::regex re(R"(^(2[0-3]|[01][0-9]):?([0-5][0-9])?:?([0-5][0-9])?$)");`
	* `static const std::regex reExt(
      R"(^(2[0-3]|[01][0-9]):?([0-5][0-9]):?([0-5][0-9])(Z|[+-](?:2[0-3]|[01][0-9])(?::?(?:[0-5][0-9]))?)$)");`

I then gave up waiting, must have been close to half an hour at least.

Note: there seem to be a few more places using regex that may or may not be affected

### PR DESCRIPTION
**Pull request by neheb**
This is mega slow on MinGW.

Should fix: https://github.com/Exiv2/exiv2/issues/2858

ping @wormnest

### CODE REVIEW

**Comment by neheb** [comment]
funny. QuickFuzz crashes but CI Fuzz does not.

**Comment by Wormnest** [comment]
I'll try to test it soon, but have some other things I need to tackle first.

**Comment by Wormnest** [comment]
@neheb Just did a test using MSYS2 UCRT64 and the test images and a few others with relevant metadata loaded fine, thanks a lot!.

I do know there are 2 other places in exiv2 that still use regex:
- In `version.cpp`, not sure how to trigger that.
- In `canonmn.cpp`. This seems to be about CanonCs LensType. Not sure if this would be triggered by using the library, or only from exiv2 directly. I found one image I had that has this tag and it did load fine in GIMP.

**Comment by neheb** [comment]
@Wormnest version.cpp only triggers with the exiv2 app I believe.

The canonmn looks like it can be triggered by the library.  That regex looks like a nightmare to remove.

Anyway, I have to see why the fuzz test fails...

**Comment by neheb** [comment]
Alright, passes now.

**Comment by kmilos** [comment]
@neheb I'm suspecting the canonmn std::regex introduction coincided w/ this regression: https://github.com/Exiv2/exiv2/issues/2746

**Comment by neheb** [comment]
Sounds about right. The option is to switch to a 3rd party regex library or write a bunch of code to get something resembling that regex. Beyond the scope of this PR.

**Comment by kmilos** [comment]
> Beyond the scope of this PR.

Sure is, just making sure we got the big picture right.

File: subprojects/libobsd.wrap
Comment (kmilos) [inline subprojects/libobsd.wrap:1]: This seems unrelated to the regex change? Maybe a separate PR please?

**Comment by neheb** [comment]
@kmilos rebased.

File: src/value.cpp [hunk 1]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
+    EXV_WARNING << Error(ErrorCode::kerUnsupportedTimeFormat) << "\n";
+#endif
+    return 1;
+  };
+
+  if (buf.size() < 2)
+    return printWarning();
+
+  for (auto c : buf)
+    if (c != ':' && c != '+' && c != '-' && !std::isdigit(c))
```
Comment (kmilos) [inline src/value.cpp:908]: Might need to include and handle 'Z' as well? https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators

Comment (kmilos) [reply src/value.cpp:908]: Also note in the wikipedia page that the minus character could be either the ASCII hyphen or the real minus code in UTF-8.

File: src/value.cpp [hunk 2]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
+      return printWarning();
+
+  size_t mpos;
+  size_t spos;
+  if (buf.find(':') != std::string::npos) {
```
Comment (kmilos) [inline src/value.cpp:913]: Might also need to guard against this corner case? `hhmmss+hh:mm`

File: src/value.cpp [hunk 3]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
+    spos = 4;
+  }
+
+  auto hi = std::stoi(buf.substr(0, 2));
+  if (hi > 24)
+    return printWarning();
+  time_.hour = hi;
+  if (buf.size() > 3) {
+    auto mi = std::stoi(buf.substr(mpos, 2));
+    if (mi > 60)
```
Comment (kmilos) [inline src/value.cpp:927]:
```suggestion
    if (mi > 59)
```

File: src/value.cpp (hunk 2 above)
Comment (kmilos) [reply src/value.cpp:913]: Also, shorter formats might omit minutes and/or seconds, but still include time zone? Maybe check if `mpos` and `spos` come before eventual '+', '-', or 'Z'?

File: src/value.cpp [hunk 4]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
+    auto posColon = format.find(':');
+    if (posColon == std::string::npos) {
+      // Extended format
+      auto tzhi = std::stoi(format.substr(0, 3));
+      if (tzhi > 24)
+        return printWarning();
+      time_.tzHour = tzhi;
+      if (format.size() > 3) {
+        int minute = std::stoi(format.substr(3));
+        if (minute > 60)
```
Comment (kmilos) [inline src/value.cpp:957]:
```suggestion
        if (minute > 59)
```

File: src/value.cpp [hunk 5]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
         time_.tzMinute = time_.tzHour < 0 ? -minute : minute;
       }
+    } else {
+      // Basic format
+      auto tzhi = std::stoi(format.substr(0, posColon));
+      if (tzhi > 24)
+        return printWarning();
+      time_.tzHour = tzhi;
+      int minute = std::stoi(format.substr(posColon + 1));
+      if (minute > 60)
```
Comment (kmilos) [inline src/value.cpp:968]:
```suggestion
      if (minute > 59)
```

File: src/value.cpp (hunk 1 above)
Comment (neheb) [reply src/value.cpp:908]: I’ll handle Z. The rest seems overkill IMO.

File: src/value.cpp (hunk 2 above)
Comment (neheb) [reply src/value.cpp:913]: Probably best to cross that line if it ever appears.

File: src/value.cpp [hunk 6]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
+    auto mi = std::stoi(buf.substr(mpos, 2));
+    if (mi > 59)
+      return printWarning();
+    time_.minute = std::stoi(buf.substr(mpos, 2));
+  } else {
+    time_.minute = 0;
+  }
+  if (buf.size() > 5) {
+    auto si = std::stoi(buf.substr(spos, 2));
+    if (si > 59)
```
Comment (kmilos) [inline src/value.cpp:935]:
```suggestion
    if (si > 60)
```

The standard actual allows for "60" to account for the "leap second".

File: src/value.cpp [hunk 7]
```diff
@@ -895,39 +894,83 @@ int TimeValue::read(const std::string& buf) {
...
+    time_.second = std::stoi(buf.substr(spos, 2));
+  } else {
+    time_.second = 0;
+  }
+
+  auto fpos = buf.find('+');
+  if (fpos == std::string::npos)
+    fpos = buf.find('-');
+
+  if (fpos != std::string::npos) {
```
Comment (kmilos) [inline src/value.cpp:946]: Are `tzHour` and `tzMinute` guaranteed to be zero-initialized? Perhaps we should, just in case? Otherwise, need to handle the 'Z' case explicitly as well if no '+' or '-' are found?

Comment (neheb) [reply src/value.cpp:946]: Beats me. This was the original code.

Comment (kmilos) [reply src/value.cpp:946]: I realise that, would've made the same comment had I reviewed it back then... Just wondering if we should take this opportunity to improve it?

Comment (kmilos) [reply src/value.cpp:946]: Looks like it's zeroed in the constructor, so should be good to go.

**Review by kmilos — APPROVED** [review]

**Comment by kmilos** [comment]
@mergify backport 0.28.x
